### ISSUE TITLE
Array conversion when adding serialized Annotation

### ISSUE THREAD

**Issue by pneves-ciandt** [stephanrauh/ngx-extended-pdf-viewer#2677]
When calling the addEditorAnnotation method on the PdfViewerService, you can pass a serialized string or the EditorAnnotation object (serializedAnnotation: string | EditorAnnotation). But the issue is when passing a string, the first thing that method does is check if it is an array, and if not it converts it into one.
When it then passes the object to the .#annotationEditorUIManager.addSerializedEditor, it checks if it is a string to deserialize, but since you added it to an array, it will never be true.
Maybe this:

```js
async addEditorAnnotation(data) {
    if (!Array.isArray(data)) {
      data = [data];
    }
    await this.#annotationEditorUIManager.addSerializedEditor(data, true, true, false);
  }
```

could be changed into:

```js
async addEditorAnnotation(data) {
    if (!Array.isArray(data) && typeof data !== 'string') {
      data = [data];
    }
    await this.#annotationEditorUIManager.addSerializedEditor(data, true, true, false);
  }
```

or something....

**Comment by stephanrauh** [comment]
Yes, you're right. However, I think it makes even more sense to convert the string to a JSON object in the calling method, because that's what happens in the private method. This way, you're allowed to forget the array brackets when calling the public method with a string.

What do you think? I'll prepare the change and wait for your feedback before publishing it.

**Comment by pneves-ciandt** [comment]
I agree. I only suggested that because some people might use the string option. I am still investigating another issue. I am trying to add a highlight annotation, but if I pass it as the type EditorAnnotation, it complains about some properties that are not in the Annotation class model (outlines for example). I tried passing it as string (I modified the js to properly deserialize), tried casting as HighlightEditorAnnotation and even just EditorAnnotation, but for some reason when it gets to deserializing into the editor, the check "if (data instanceof HighlightAnnotationElement)" always fails, so it gets handled as a generic annotation without the outlines and everything else. And after that it breaks the annotations again (similar to the issue I had before upgrading to angular 17). I am trying to debug and find out why is that and what can I change to make this work....

**Comment by stephanrauh** [comment]
Awesome. Your improvement is scheduled to be part of version 22.0.1!

As for the highlight annotation: as far as I understand, the API I'm using simply doesn't support adding highlights. The rationale is that it doesn't makes sense to copy and paste highlights, so the paste API doesn't support them, and I'm highjacking the paste API. However, I'm not 100% sure. I've received several bug reports that make it sound like my implementation does have an impact, it just doesn't deliver the result it promises.

I guess I should either throw an exception if you're trying to add a highlight, or I should go the extra mile and implement the feature. Unfortunately, I don't have a clue how much work that might be. Maybe you can help me?

**Comment by pneves-ciandt** [comment]
Ah, I see. I'll take a look at the code and see if I have any ideas! I'll make a PR if I come up with something

**Comment by stephanrauh** [comment]
As for the PR: have a look at the discussion over there. Chances are the pdf.js team have already solved the bug, and I'm just a bit late when it comes to merge their changes to ngx-extended-pdf-viewer.

**Comment by stephanrauh** [comment]
Your array conversion improvement has landed with version 22.0.1.

Enjoy!
Stephan
